### ISSUE TITLE
Some use of reflection will not work anymore in Android P

### ISSUE THREAD

**Issue by netomi** [stefan-zobel/streamsupport#2]
Android P reports various warnings when using the streamsupport library:

zygote  : Accessing hidden field Lsun/misc/Unsafe;->theUnsafe:Lsun/misc/Unsafe; (light greylist, reflection)
zygote  : Accessing hidden field Ljava/lang/Integer;->value:I (light greylist, reflection)
zygote  : Accessing hidden field Ljava/lang/Thread;->contextClassLoader:Ljava/lang/ClassLoader; (dark greylist, reflection)
zygote  : Accessing hidden field Ljava/util/concurrent/atomic/AtomicInteger;->value:I (dark greylist, reflection)

The problematic classes are TLRandom which accesses the contextClassLoader and Striped64 which accesses a private field in the AtomicInteger class. There might be other classes as well as our test cases do not cover everything of the streamsupport library.

Basically, warnings wrt light greylist are merely a warning for now, but might be restricted in later versions. Warnings wrt dark greylist will most likely be restricted in Android P and will not work anymore.

See also https://developer.android.com/preview/restrictions-non-sdk-interfaces.html for a more detailed explanation.

**Comment by stefan-zobel** [comment]
Hi Thomas @netomi

thanks for letting me know. Sigh, that's bad news (streamsupport currently runs even on Java 10 / 11 with only a single warning - and that is caused by a test, not the library itself).

It's especially annoying since absolutely nothing has changed at those critical places in the Android P master until now.

There are lot more usages of Unsafe than the few that you mentioned. I surely could get rid of some of them, but unlikely all of them (as this would cripple the functionality especially for parallel use cases).

I'll have to let sink that in and think about it. Also, seems I need an additional device for Android P testing now ;)

One possible approach could be to fork streamsupport a second time (android-retrostreams being the first fork), this time creating a version that is targeted especially at Android that shares no code with the JVM version in those critical areas.
As I said, I have to think about it. However, nothing will happen during the next two weeks (I'm away on vacation).

Pertaining to DexGuard: that's simply great! May I ask why you've decided to use streamsupport instead of android-retrostreams?

Regards,
Stefan

**Comment by stefan-zobel** [comment]
Hi Thomas,

working backwards ...

disabling delegation / system property: yeah, that's right, it's not Android style. I've initially used that approach on Java 8 and simply adopted it when I introduced delegation for Android 7.
What do you mean by "hard-coded flag"? A "static final" something that you'd set by instrumentation? Can you give me an example?

As to why you've chosen streamsupport - I see, makes sense. Android tools - same here ;)

The places where Unsafe or any other kind of reflection is hard to replace have mostly to do with fork/join (parallel streams).
I've always wondered how important that use case is on Android devices. Do you have any data points (or an opinion) about this?

Regards,
Stefan

**Comment by netomi** [comment]
Hi Stefan,

for the delegation issue, I had something like this in mind:

static boolean DELEGATION_ENABLED_VIA_INSTRUMENTATION = true;
static final boolean DELEGATION_ENABLED = getBooleanPropVal(DELEGATION_ENABLED_P, true) && DELEGATION_ENABLED_VIA_INSTRUMENTATION;

It is important the the field DELEGATION_ENABLED_VIA_INSTRUMENTATION is not final as otherwise the compiler would inline its value.

But after reading the Known_Problems.txt and looking at the relevant source code, I wonder if this is even necessary anymore. From Spliterators.java:

if (HAS_STREAMS && (DELEGATION_ENABLED || IS_JAVA9) && !hasAndroid7LHMBug(c)) {
      // always use spliterator delegation on Java 9 from 1.5.6 onwards
      // https://sourceforge.net/p/streamsupport/tickets/299/
      return delegatingSpliterator(c);
}

Does this mean that delegation is disabled by default for API 24 & 25 for LinkedHashMaps?
Probably the Known_Problems.txt has to be updated as it is mentioned that delegation is enabled by default and has to be disabled on Android 24 & 25 to avoid the bug (or use a slightly modified code).

**Comment by stefan-zobel** [comment]
Hi Thomas,

I've introduced that additional LHM check measure as a precaution for users that don't read (or understand) this note.

Yes, LinkedHashMaps are exempt from Spliterator delegation on API 24/25 even if Spliterator delegation is enabled generally.
However this works only insofar as it can be detected at runtime that the Collection views come from a LHM. For example, wrap a LHM in a Collections.UnmodifiableMap and you are screwed. If you want to guard against such cases you either have to disable Spliterator delegation altogether or manually program around this Android 7.x bug.

That's what I wanted to express in this note. Obviously, I didn't fully succeed.

As for the DELEGATION_ENABLED_VIA_INSTRUMENTATION switch - I have no problems with that as long as it can be package private.

I've thought about that shortcoming myself but I didn't want to introduce anything in the public API that hasn't its equivalent in the Java 8 API (and that most users wouldn't understand anyways).

Regards,
Stefan

**Comment by netomi** [comment]
Hi Stefan,

maybe I was too fast reading the file. After your explanation and reading it again it is clear to me now and it does not need any further change imho.

For the additional field: it can even be private, just not final. So far we do not yet try to disable delegation but it might be interesting for some users that want a fool-proof method to avoid known bugs without changing their code. Actually I never tried to set a system property in an Android application, but it might work as well, as long as it is done before any streamsupport class is accessed. I should create a separate issue explaining the use-case, but it would only be for convenience.

**Comment by stefan-zobel** [comment]
Yeah, the "not final" bit is clear :)

You are right, setting the system property works on Android as you say. That's how I'm doing it in my Android tests.

What I don't quite understand is why you want to disable Spliterator delegation at all. If you disable it and use a Collection that employs one of my "native specializations" you run into even more risks the newer the Android version happens to be.
That was one of the main reasons why I've introduced delegation on Android in the first place.

Perhaps this one helps to understand the history: https://github.com/streamsupport/streamsupport/pull/7

Regards,
Stefan

**Comment by netomi** [comment]
Hi Stefan,

ok thanks for the clarification. I was not aware of the fact that the non-native spliterators are custom made and use reflection. I did a quick check in the source code and assumed that the original Oracle implementation is used, but you explained that this is broken on Android.

Considering this, the flag to disable the delegation is indeed not very useful anymore and it is better to include the specific warning about the use of LHM spliterators on Android 24 & 25.

**Comment by stefan-zobel** [comment]
Hi Thomas,

the "native" spliterators must be custom since they don't exist on Java 6/7 or API <= 23.
They **are** identical to the Oracle implementation, modulo the fact that they **have** to use reflection since the Oracle Spliterators also access private class internals (which they obviously can do and I can't do without reflection).

The IteratorSpliterator alternative has its own problems. While it works quite well in general getting the Spliterator characteristics right in all cases would be a real challenge.

The decision to remove the "java8.util.Spliterators.assume.oracle.collections.impl=false" flag (for Android) in 1.3.1-stable was quite intentional.
If I had maintained that flag I'd have to test / maintain even more possible combinations than I have today.

Regards,
Stefan

**Comment by netomi** [comment]
Hi Stefan,

I did not realize that you need reflection to perform the same functionality but it makes perfect sense now. The problems with disabling delegation would only be made worse with the announced changes in Android P so it is not really an option anymore. Thanks for the clarification.

**Comment by netomi** [comment]
Sounds great, have a nice vacation!

**Comment by stefan-zobel** [comment]
Hi Thomas,

I've only done a very superficial short test. The two dark greylist warnings (contextClassLoader and AtomicInteger) are probably gone.

I've actually used android-retrostreams for the test but that should make no difference. You can try it yourself with the attached snapshot if you want. A more thorough probe will follow when I'm back.

Regards,
Stefan

[streamsupport-1.7.0-SNAPSHOT.jar.zip](https://github.com/stefan-zobel/streamsupport/files/1929910/streamsupport-1.7.0-SNAPSHOT.jar.zip)

**Comment by stefan-zobel** [comment]
The two most critical issues (for now) are the "dark" greylist warnings.

The CCL issue relates to https://sourceforge.net/p/streamsupport/tickets/275/ and https://bugs.openjdk.java.net/browse/JDK-8172726
That feature is for containerized environments (Servlet Engines and the like) and doesn't have to work on Android (and actually never did work on API >= 24). 
So, I've simply dropped this.

The AtomicInteger issue is caused by my ForkJoinPool.MemBar emulation of the sun.misc.Unsafe fence() methods. I have replaced AtomicInteger by my own "Mock" class.

**Comment by netomi** [comment]
Hi Stefan,

using the version 1.7.0-SNAPSHOT, I get the following warnings:

04-20 10:42:24.684 17294 17294 W zygote  : Accessing hidden field Lsun/misc/Unsafe;->theUnsafe:Lsun/misc/Unsafe; (light greylist, reflection)
04-20 10:42:24.687 17294 17294 W zygote  : Accessing hidden field Ljava/lang/Integer;->value:I (light greylist, reflection)

They are in the light greylist so it will hopefully be ok for Android P. Our tests only cover a small subset of the whole library though.

**Comment by stefan-zobel** [comment]
Hi Thomas,

thanks for the confirmation.

The streamsupport test suite is quite comprehensive. In the first week of May i'll have the time and resources for a complete run on all platforms.
I'll approach the Integer.value issue subsequently. For `Unsafe` I can't do anything without kicking the whole parallelism out of the library.

Other than that we have to wait for any further restrictions Google may come up with in the follwowing previews.

Regards,
Stefan

**Comment by netomi** [comment]
Hi Stefan,

for our product we have a test suite that executes tests directly on a host dalvikvm, I can also give that a try, as it should be much easier to execute them. I have not yet built a dalvikvm for Android P but it was anyway on my TODOs.

**Comment by stefan-zobel** [comment]
Hi Thomas @netomi

The light greylist "hidden field Ljava/lang/Integer;->value:I" warning should be gone now.
I've only done a quick'n dirty run of the streamsupport test suite on the Android P emulator using android-retrostreams.

I'll do a thorough test later this week.

Regards,
Stefan

[streamsupport-1.7.0-SNAPSHOT-2.zip](https://github.com/stefan-zobel/streamsupport/files/1964257/streamsupport-1.7.0-SNAPSHOT-2.zip)
[streamsupport-atomic-1.7.0-SNAPSHOT-2.zip](https://github.com/stefan-zobel/streamsupport/files/1964258/streamsupport-atomic-1.7.0-SNAPSHOT-2.zip)

**Comment by stefan-zobel** [comment]
Hi Thomas @netomi

The streamsupport test suite passes all tests on all test platforms. In addition, I've run it on the current Android P preview rev. 3 image (everything ok there too).
I'm pondering an "interim" release of the current state of things as 1.6.3-stable.

However, I'd suggest to leave this issue open until we can test on a final Android P release candidate.

Regards,
Stefan

**Comment by paulduffin** [comment]
Hi, I work on the Android development team (work with @15characterlimi who did some work to support delegating spliterators on Android). Apologies for reopening an old issue but it seemed an appropriate place to comment.

The reason that you no longer see the warnings about the use of internal APIs is probably because by default the early releases of P logged that information but the released version of P does not. The internal fields mentioned in the initial comment on this thread are still part of what is called the "light grey list" and still at risk of being removed at a later date.

I am currently looking at APKs that use reflection to access the `Integer.value` field and this library is a major source of those uses. I've had a look at the code and it seems as though it is only used in `TLRandom` in order to allow the `int getInitializedProbe(Integer)` method return what is essentially a boolean flag by updating the supplied `Integer`.

Is there any specific reason why it has to be done that way or would one of the following approaches be usable?

1. Change to return a long and encode the (boolean, int) flag in the long, e.g. by casting the int to long (treating it as unsigned) and then encoding the boolean as the sign, e.g. -1234 would mean that it was uncontended, and 1234 would mean that it was not uncontended.
2. Use AtomicBoolean.
3. Use a custom class.

The reason we are making these changes is to try and improve app compatibility from one release to the next. If there is no way for you to achieve what you want without accessing these internal fields/methods then please could you raise a feature request (described in the link in the first comment) so that we can update the API to support your use case.

Many Thanks, Paul Duffin

**Comment by paulduffin** [comment]
I was looking at the source in GitHub which only send to go to 1.6.0.

On Nov 2, 2018 20:12, "stefan-zobel" <notifications@github.com> wrote:

Hi Paul,

which version of streamsupport are you looking at? I thought I had removed
that method in commit
598931a#diff-d924a4f14b8a00eed9e2b6fef528e677
<https://github.com/stefan-zobel/streamsupport/commit/598931ad2559f9f431b75daefcbeb523cfe4e5ee#diff-d924a4f14b8a00eed9e2b6fef528e677>

It shouldn't exist anymore in streamsupport >= 1.6.3.
Please let me know if you find any further compatibility issues.

Regards,
Stefan

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/stefan-zobel/streamsupport/issues/2#issuecomment-435484072>,
or mute the thread

**Comment by stefan-zobel** [comment]
Oh, you mean https://github.com/streamsupport/streamsupport

Wrong repository, I'm sorry. That one is a mirror of my old sourceforge hg repo not operated by me.
The gentleman that started this mirror kindly ignored my update requests. The up-to-date source is now here.

Kind regards,
Stefan

**Comment by paulduffin** [comment]
Ah, that explains it. I did wonder why there was two of them. Apologies for
wasting your time.

**Comment by stefan-zobel** [comment]
Never mind! Please keep me updated whenever you find a new compatibility issue.

Have a nice weekend,
Stefan

**Comment by netomi** [comment]
Hi,

the manual includes a section, Gradle plugin | Java 8 stream support.
Additionally there are also samples like StreamSupport, TimeAPI that illustrate the use. For proguard you have to look at the latest beta release.

**Comment by stefan-zobel** [comment]
Hi Zhou Jian @zhoujian365

As a start, I'd try adding the streamsupport jar via the "-injar" parameter and set the "-target" to 1.6 or 1.7
That worked for me in a quick test I did back in November. See also the discussion in https://github.com/stefan-zobel/streamsupport/issues/5

HTH,
Stefan

**Comment by zhoujian365** [comment]
Hello,

I notice that the Proguard Gradle Plugin requires com.android.application plugin, but I need to use it in a library project.

I’ve read the code of the plugin and it checks com.android.application plugin at the beginning. Is it possible to modify this to use it in a library project?

Thanks.
On 01/24/2019 13:31，Thomas Neidhart<notifications@github.com><mailto:notifications@github.com> wrote：

Hi,

the manual includes a section, Gradle plugin | Java 8 stream support.
Additionally there are also samples like StreamSupport, TimeAPI that illustrate the use. For proguard you have to look at the latest beta release.

**Comment by netomi** [comment]
Hi,

thanks for the feedback. We did not think about this use-case so far, also I am not fully aware if adding transform tasks (the plugin is configuring and adding a transform task) in library projects is supported.

In any case, it should be possible to apply ProGuard as a gradle task, but it might not be fully integrated into the way the Android plugin handles library modules.

**Comment by zhoujian365** [comment]
Hi Neidhart,

I’ve figured how to transform and backport Java 8 features in library projects. I’ll be able to verify it today or tomorrow. How can I make a merge request back to your project?

<https://maas.mail.163.com/dashi-web-extend/html/proSignature.html?ftlId=3&name=%E5%91%A8%E5%89%91&uid=zhoujian365%40outlook.com&iconUrl=https%3A%2F%2Fmail-online.nosdn.127.net%2Fwzpmmc%2F35d3a44f7358ef88868c9b7337fc6cca.jpg&company=%E9%98%BF%E9%87%8C%E5%B7%B4%E5%B7%B4&position=%E9%AB%98%E7%BA%A7%E6%97%A0%E7%BA%BF%E5%BC%80%E5%8F%91%E5%B7%A5%E7%A8%8B%E5%B8%88&items=%5B%2218662268173%22%2C%22zhoujian365%40outlook.com%22%5D>
[https://mail-online.nosdn.127.net/wzpmmc/35d3a44f7358ef88868c9b7337fc6cca.jpg]
周剑
Alibaba  -  S
Senior Engineer
zhoujian365@outlook.com

签名由 网易邮箱大师<https://mail.163.com/dashi/dlpro.html?from=mail81> 定制

On 02/18/2019 15:51，Thomas Neidhart<notifications@github.com><mailto:notifications@github.com> wrote：

Hi,

thanks for the feedback. We did not think about this use-case so far, also I am not fully aware if adding transform tasks (the plugin is configuring and adding a transform task) in library projects is supported.

In any case, it should be possible to apply ProGuard as a gradle task, but it might not be fully integrated into the way the Android plugin handles library modules.

**Comment by netomi** [comment]
Hi,

you can add a feature request here: https://sourceforge.net/projects/proguard/
If you can share your solution we can integrate it for a next release.

**Comment by E3V3A** [comment]
@zhoujian365  How did you do it?
